vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/model/TestRunResultsAccumulator.kt

```diff
@@ -0,0 +1,196 @@
+package com.malinskiy.marathon.android.model
+
+import com.malinskiy.marathon.log.MarathonLogging
+import java.util.*
+
+/**
+ * Holds results from a single test run.
+ *
+ *
+ * Maintains an accurate count of tests, and tracks incomplete tests.
+ *
+ *
+ * Not thread safe! The test* callbacks must be called in order
+ */
+class TestRunResultsAccumulator {
+
+    val logger = MarathonLogging.logger { }
+
+    var name: String = "not started"
+        private set
+    val testResults = LinkedHashMap<TestIdentifier, AndroidTestResult>()
+    private val runMetrics = HashMap<String, String>()
+    var isRunComplete = false
+    var isCountDirty = false
+    var elapsedTime: Long = 0
+        private set
+
+    private var statusCounts: Map<AndroidTestStatus, Int> = mutableMapOf()
+
+    /**
+     * Return the run failure error message, `null` if run did not fail.
+     */
+    var runFailureMessage: String? = null
+        private set
+
+    var aggregateMetrics = false
+
+    val completedTests: Set<TestIdentifier>
+        get() {
+            val completedTests = LinkedHashSet<TestIdentifier>()
```

Comment:
Do we care about insertion order? If not, then why we need LinkedHashSet here?

Reply:
The original implementation of ddmlib [cares about the order](https://android.googlesource.com/platform/tools/base/+/tools_r22/ddmlib/src/main/java/com/android/ddmlib/testrunner/TestRunResult.java?autodive=0%2F%2F#38)

I think it's because some libs might do diff on the result.